OpenShift Integration: support for `oc new-app`

dustymabe opened the pull request:
This PR is an initial attempt at supporting bootstrapping Atomic App 
embedded applications view the `oc new-app` command in OpenShift.

aweiteka commented:
LGTM

dustymabe commented:
Note I also updated the provider documentation. It is in a separate PR: #455.

dustymabe commented:
@rtnpro @cdrage.. This is a big feature for us to add. The more eyes the better on this one.

cdrage commented on atomicapp/cli/main.py:
```diff
@@ -250,6 +251,13 @@ def run(self):
         cmdline = sys.argv[1:]  # Grab args from cmdline
 
+        # If we are running in an openshift pod (via `oc new-app`)
+        # then there is no cmdline but we want to default to "run". In
```
For us non-openshift folks what does 'run' mean? as in, `oc run`?

dustymabe replied:
No. I mean "atomicapp run". I'll try to update to make it more clear

cdrage commented on atomicapp/constants.py:
```diff
@@ -47,9 +47,11 @@
 HOST_DIR = "/host"
 
+# Use /tmp/ so that we can grab the lock as non-root
+LOCK_FILE = "/tmp/atomicapp.lock"
```
:+1:

dustymabe replied:
Did some more thinking about this. I have to change it back and solve it a different way. In the past multiple instances of atomicapp running on a system share /run/lock on the host because /run is mounted into the container. So they would be competing for the same lockfile. 

In this case they would not because /tmp from the host isn't mounted in the container.

cdrage commented on atomicapp/nulecule/main.py:
```diff
@@ -90,6 +90,16 @@ def __init__(self, app_spec, destination=None, answers_file=None):
             if os.path.isfile(os.path.join(self.app_path, ANSWERS_FILE)):
                 self.answers_file = os.path.join(self.app_path, ANSWERS_FILE)
 
+        # TODO: put this in a better place in the future.
```
Better place could actually just be the `openshift.py` under providers.

I've been thinking lately of refactoring how we use each provider since we're adding new commands in the future. For example, changing `def undeploy` to `def stop` and `def deploy` to `def run`. Making it a bit more modular for new providers to be added. 

Perhaps a `def check_host` or something similar could help

dustymabe replied:
The problem is that we don't call the provider code until we know what provider we are using. In this case we actually set that information here.

cdrage replied:
Yeah, let's change this in another PR, no effects, it's just semantics.

cdrage replied:
@dustymabe Okay, I'm okay for merging, but this needs to be changed in the future as I feel as though it's not suppose to belong there.

dustymabe replied:
Right.. I feel like if we add support to kubernetes for this same thing then we'll come up with the right place to "detect" this stuff and build out a framework for it.

Actually I'd like a single place where answers.conf + CLI args + defaults + detected environment overrides are all merged into one. Maybe I'll open an issue for that.

cdrage commented on atomicapp/utils.py:
```diff
@@ -114,6 +115,45 @@ def __init__(self, target_path, workdir=None):
+        return False
+
+    @staticmethod
+    def get_openshift_api_endpoint_from_env():
```
sorry about being nitpicky, but other function calls within Utils are camelCase.

@rtnpro i think you brought this up previously before, what does it say in pep-8 specifically? otherwise, we should change utils to conform to how rtnpro did his refactor and made everything snake_case

kadel replied:
I believe that pep8 says that function names should be snake_case and classes CamelCase.

dustymabe replied:
So I should leave this snake case and we'll have another PR that makes everything in here snake_case ?

cdrage replied:
Yeah, it's just semantics, we'll update that in another PR.

dustymabe commented:
Re-worked the commits a bit. Fixed lockfile issue in a different way. Added comments about defaulting to "atomicapp run". 

@cdrage, please review.

cdrage commented:
@dustymabe 
reviewed, just one comment, otherwise, LGTM :+1:
